### PR TITLE
Allow to perform "metadata-only" ALTER of partition key.

### LINKED ISSUE #7513: Changes of Enum in partition key don't work
From #2667. thanks @develar  

```
CREATE TABLE report
(
    `product` Enum8('IU' = 1, 'WS' = 2), 
    `machine` String, 
    `branch` String, 
    `build_c1` String, 
    `build_c2` String, 
    `build_c3` String, 
    `build_time` DateTime, 
    `generated_time` DateTime
)
ENGINE = MergeTree
PARTITION BY (product, toYYYYMM(generated_time))
ORDER BY (product, machine, branch, build_c1, build_c2, build_c3, build_time, generated_time)
SETTINGS old_parts_lifetime = 10;

alter table report modify column product Enum8('IU' = 1, 'WS' = 2, 'PS' = 3);
```

Got:
```
Received exception from server (version 19.16.1):
Code: 44. DB::Exception: Received from localhost:9000. DB::Exception: Trying to ALTER key column product. Stack trace:

0. 0x55e1ab472d20 StackTrace::StackTrace() /build/obj-x86_64-linux-gnu/../dbms/src/Common/StackTrace.cpp:228
1. 0x55e1ab472af5 DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int) /build/obj-x86_64-linux-gnu/../dbms/src/Common/Exception.h:27
2. 0x55e1ab1ff52d DB::MergeTreeData::checkAlter(DB::AlterCommands const&, DB::Context const&) [clone .cold] /usr/include/c++/9/bits/basic_string.h:222
3. 0x55e1aea283af DB::StorageMergeTree::alter(DB::AlterCommands const&, DB::Context const&, DB::TableStructureWriteLockHolder&) /build/obj-x86_64-linux-gnu/../dbms/src/Storages/StorageMergeTree.cpp:273
4. 0x55e1aef5ab33 DB::InterpreterAlterQuery::execute() /usr/include/c++/9/bits/shared_ptr_base.h:729
5. 0x55e1ae917916 DB::executeQueryImpl(char const*, char const*, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, DB::ReadBuffer*, bool) /build/obj-x86_64-linux-gnu/../dbms/src/Interpreters/executeQuery.cpp:314
6. 0x55e1ae918ade DB::executeQuery(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, bool) /build/obj-x86_64-linux-gnu/../dbms/src/Interpreters/executeQuery.cpp:562
7. 0x55e1ab4fae74 DB::TCPHandler::runImpl() /build/obj-x86_64-linux-gnu/../dbms/programs/server/TCPHandler.cpp:247
8. 0x55e1ab4fb8ab DB::TCPHandler::run() /build/obj-x86_64-linux-gnu/../dbms/programs/server/TCPHandler.cpp:1225
9. 0x55e1af392970 Poco::Net::TCPServerConnection::start() /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerConnection.cpp:57
10. 0x55e1af39308d Poco::Net::TCPServerDispatcher::run() /usr/include/c++/9/bits/atomic_base.h:327
11. 0x55e1b0a69261 Poco::PooledThread::run() /usr/include/x86_64-linux-gnu/c++/9/bits/gthr-default.h:748
12. 0x55e1b0a6700c Poco::ThreadImpl::runnableEntry(void*) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/include/Poco/AutoPtr.h:205
13. 0x55e1b11d8ba0 execute_native_thread_routine /usr/lib/debug/usr/bin/clickhouse
14. 0x7f67701416db start_thread /lib/x86_64-linux-gnu/libpthread-2.27.so
15. 0x7f676fa5e88f clone /build/glibc-OTsEL5/glibc-2.27/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:97
0 rows in set. Elapsed: 0.147 sec. 
```

metadata only changes of ORDER BY is allowed:
https://github.com/ClickHouse/ClickHouse/commit/9918965c
https://github.com/yandex/ClickHouse/blob/master/dbms/tests/queries/0_stateless/00427_alter_primary_key.sh

But not for partition key

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow to perform "metadata-only" ALTER of partition key such as adding more elements to Enum data type. This fixes #7513.

Details:
This feature was discussed on ClickHouse Meetup online on Jun 18.

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -1219,18 +1220,77 @@ bool isMetadataOnlyConversion(const IDataType * from, const IDataType * to)
...
+            { typeid(DataTypeEnum8),    typeid(DataTypeEnum16) },
+            { typeid(DataTypeEnum8),    typeid(DataTypeInt8)   },
+            { typeid(DataTypeEnum8),    typeid(DataTypeInt16)  },
+            { typeid(DataTypeEnum8),    typeid(DataTypeInt32)  },
+            { typeid(DataTypeEnum8),    typeid(DataTypeInt64)  },
+
+            { typeid(DataTypeEnum16),   typeid(DataTypeEnum16) },
+            { typeid(DataTypeEnum16),   typeid(DataTypeInt16)  },
+            { typeid(DataTypeEnum16),   typeid(DataTypeInt32)  },
+            { typeid(DataTypeEnum16),   typeid(DataTypeInt64)  },
```
Comment (filimonov) [inline src/Storages/MergeTree/MergeTreeData.cpp:1249]: Extendenting the size of column (like from UInt8 to uint64), will also require refreshing `minmax_` file. Is it recounted now on ALTERs automatically?

File: tests/queries/0_stateless/01344_alter_enum_partition_key.sql [hunk 2]
```diff
@@ -0,0 +1,54 @@
+DROP TABLE IF EXISTS test;
```
Comment (alesapin) [inline tests/queries/0_stateless/01344_alter_enum_partition_key.sql:1]: Need a test for Replicated table.

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 1 above)
Comment (alesapin) [reply src/Storages/MergeTree/MergeTreeData.cpp:1249]: Yes, seems like we will not be able to deserialize partition key minmax from disk https://github.com/clickhouse/ClickHouse/blob/master/src/Storages/MergeTree/IMergeTreeDataPart.cpp#L44-L55. Also we have a separate method for internal structures update https://github.com/clickhouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeData.cpp#L418.

**Comment by alexey-milovidov** [comment]
> Integration tests

Broken in #11903

**Review by alesapin — APPROVED** [review]
LGTM. Also, we have a check for partition key in ReplicatedMergeTreeMetadata, but it verifies only column names, without types.

**Pull request merged**